jerry-core/ecma/base/ecma-helpers-value.c

```diff
@@ -275,21 +300,151 @@ ecma_check_value_type_is_spec_defined (ecma_value_t value) /**< ecma value */
 inline ecma_value_t __attr_const___ __attr_always_inline___
 ecma_make_simple_value (const ecma_simple_value_t simple_value) /**< simple value */
 {
-  return (((ecma_value_t) (simple_value)) << ECMA_VALUE_SHIFT) | ECMA_TYPE_SIMPLE;
+  ecma_value_t flags = ECMA_SIMPLE_VALUE_FLAG | ECMA_TYPE_NON_PTR_VALUE;
+  return (((ecma_value_t) (simple_value)) << ECMA_INTEGER_NUMBER_SHIFT) | flags;
 } /* ecma_make_simple_value */
 
 /**
- * Number value constructor
+ * Encode an integer number into an ecma-value without allocating memory
+ *
+ * Note:
+ *   The value must fit into the range of allowed ecma integer values
+ *
+ * @return ecma-value
  */
-ecma_value_t __attr_const___
-ecma_make_number_value (const ecma_number_t *ecma_num_p) /**< number to reference in value */
+inline ecma_value_t __attr_const___ __attr_always_inline___
+ecma_make_integer_value (ecma_value_integer_t integer_value) /**< integer number to be encoded */
 {
-  JERRY_ASSERT (ecma_num_p != NULL);
+  JERRY_ASSERT (ECMA_INTEGER_NUMBER_MIN <= integer_value
+                && integer_value <= ECMA_INTEGER_NUMBER_MAX);
+
+  return (ecma_value_t) ((integer_value << ECMA_INTEGER_NUMBER_SHIFT) | ECMA_TYPE_NON_PTR_VALUE);
+} /* ecma_make_integer_value */
+
+/**
+ * Allocate and initialize a new float number without checks.
+ *
+ * @return ecma-value
+ */
+static ecma_value_t __attr_const___
+ecma_create_float_number (ecma_number_t ecma_num) /**< value of the float number */
+{
+  ecma_number_t *ecma_num_p = ecma_alloc_number ();
+
+  *ecma_num_p = ecma_num;
+
+  return ecma_pointer_to_ecma_value (ecma_num_p) | ECMA_TYPE_FLOAT_NUMBER;
+} /* ecma_create_float_number */
+
+/**
+ * Create a new float number (usually from NaN).
+ *
+ * @return ecma-value
+ */
+inline ecma_value_t __attr_always_inline___
+ecma_make_float_number_value (ecma_number_t ecma_num) /**< value of the float number */
+{
+  JERRY_ASSERT (((ecma_number_t) (ecma_value_integer_t) ecma_num) != ecma_num);
+
+  return ecma_create_float_number (ecma_num);
+} /* ecma_make_float_number_value */
+
+/**
+ * Checks whether the passed number is +0.0
+ *
+ * @return true, if it is +0.0, false otherwise
+ */
+static inline bool __attr_const___ __attr_always_inline___
+ecma_is_number_equals_to_positive_zero (ecma_number_t ecma_num) /**< number */
```

Comment:
Checking negative zero is difficult, because you need to do two comparisons:

uint32_t values[2]
memcpy(values, &double, 8);

negative zero check (little-endian):
  if (values[0] == 0 && values[0] == 0x80000000)

positive zero check:
 if ((values[0] | values[1]) == 0) // only one ARM instruction + branch: orrs low, high

Furthermore converting double to two uint32 values is costly (FPU->Integer core data copy) so I would only do it when absolutely necessary (the converted value is zero).